New readme template (#2)
tforster/git-new: README.md

Inspired by Eric Meyers CSS Reset, Git New is a simple tool to enforce a consistent baseline Git experience for new projects.

> - [Git New](#git-new)

Reviewer: Is the top level heading necessary?  Is it auto-generated?

Author: Are you referring to the ToC top level? Yes, auto-generated. It's the challenge of applying web style headings to a document style MD file. Another decision we need to make. In a document, you can have multiple heading 1's. Each one corresponding to a chapter (or similar). Mostly because a non-web document has pages, whereas a web page is, well, a page. Some of my MDs can be quite long, especially architecture, etc. I would prefer to use multiple H1s in an MD, treating it like a document that just happens to be web-renderable.

Thoughts?

Reviewer: I am cool with multiple heading 1's on the page. It makes semantic sense, the same way you can have multiple H1s in HTML5 as long as they are in appropriate regions.

>   - [For Project Developers](#for-project-developers)

Reviewer: The headings / anchors here seem to be inconsistent with the content below.

Author: Hmm, should be auto-update on save. I will confirm.

    - [Prerequisites](#prerequisites)
    - [Setup and Configuration](#setup-and-configuration)
  - [For Installers](#for-installers)
    - [Prerequisites](#prerequisites-1)
    - [Setup and Configuration](#setup-and-configuration-1)
  - [Change log](#change-log)
  - [Meta](#meta)
  - [Contributing](#contributing)

## For Developers

Follow this section if you are developing with, installing or deploying this repository.

### Prerequisites

> - A good code editor. We suggest [VSCode](https://code.visualstudio.com/), [Atom](https://atom.io/) or [SublimeText](https://www.sublimetext.com/).

Reviewer: Although I like promoting good tools, is the IDE necessary for the README? The project can be run, edited, maintained, built, deployed, etc. all without the IDEs listed. Despite the incredible usefulness of the tools, I wouldn't say they are necessarily required the way Node or Git is. Thoughts?

Author: I dunno why I keep coming back to this. Yes, let's pull out the IDEs and editors.
```suggestion
- A good code editor.
```

> - [NodeJS LTS+ and NPM 6.14+](https://nodejs.org/en/download/)

Reviewer: Should we specify versions for things like Node to future proof it? I don't know if patch or even minor is necessary, but I would think major might be handy, as things could change down the road. Thoughts?

Author: What's your view on the LTS convention? When I work (assuming I am not constrained by AWS Lambda runtimes) I tend to go for the latest Node version. And I will live with the consequences of being on the breaking edge. But for collab/shared source should we stick with latest and just list the version? Or favour stability and recommend LTS?

Reviewer: I like favouring the current LTS for shared source or client projects.  However, simply listing LTS doesn't help when we try to run the project a few years later as there could be a different LTS version.  In the future, if I knew it was a specific major (even minor) version, I could use NVM to switch before running it.  This mostly applies to projects that come and go quickly and don't see long term development, where the code base would evolve as the tools do.

Author: However, if we pin to a version (use Node as an example) and someone clones the repo 2 years later does that imply they have to downgrade their version to one from 2 years prior? 

I was adding a plus sign (e.g. NPM 6.14+) but that gets a bit redundant. 

```suggestion
### Prerequisites

The versions listed for these prerequisites are current at the time of writing. More recent versions will likely work but "your mileage may vary".

- A good code editor. 
- [Node.js v12.13.0 and NPM 6.14.5](https://nodejs.org/en/download/)
```

Author: FYI, I'm using ctrl-g in GitHub to insert suggested changes that can be accepted and merged from this page (it's pretty cool). But in the above one I notice it is not replacing the full text, only the line :(

Reviewer: I think it depends on what changes over time, which is out of our control.  Pinning to a major or even minor version may guarantee that it will run on that version, where future major versions may intro breaking changes that is incompatible with our code.  Most of our projects will be fine running in the future, but there is a chance that it doesn't run on Node 20.  I think adding that line that you did makes the most sense.

Reviewer: I see what you mean, it duplicated the content in the commit.

# 4. Edit this README.md and package.json to suit your project
#    - Edit the title to match your project
#    - Edit the description to describe your project

# 5. Edit package.json
#    - Update the title to match the title of this README
#    - Update the description to match the description of this README
#    - Edit the semantic version to match your project requirements

# 6. Delete any unwanted files in /docker (or the entire /docker folder if you're not containerising)

# 7. Re-initialise .git
git init

# 8. Add your own repo's origin
git remote add origin {my-git-origin}

# 9. Update package.json to reflect the new origin and clean up any other properties by following the prompts
npm init

# 10. Delete this instruction block

# 11. Stage and commit files
git add -A
git commit -m"Initial creation"

# 12. Push
git push -u origin master

# 13. Create a develop branch (workflow dependent)
git checkout -b develop
git push -u origin develop

# Your project and repo are ready for your first feature branch
git checkout -b {my-feature-branch-name}
```

## For Project Developers

Follow this section if you will be developing new features and fixing bugs for this repository.

> ### Prerequisites

Reviewer: See notes above in #prerequisites section.

- A good code editor. We suggest [VSCode](https://code.visualstudio.com/), [Atom](https://atom.io/) or [SublimeText](https://www.sublimetext.com/).
- [NodeJS LTS+ and NPM 6.14+](https://nodejs.org/en/download/)
- [Git 2.25+](https://git-scm.com/downloads)

### Setup and Configuration

```sh
# 1. Clone this repository as your new project
git clone git@github.com:tforster/git-new.git /my/path/to/my/project-folder/
cd /my/path/to/my/project-folder/git-new

# 3. Create a new branch
git checkout -b {new-branch-name}

# 2. Make changes by editing files and/or using NPM to update dependent packages as well as package.json

# 3. Stage changes
git add -A

# 4. Commit changes
git commit -t"{detailed-commit-message}"

# 5. Push changes
git push origin {new-branch-name}

# 6. Create a pull request
```

> ## Change log

Reviewer: This should be title case, ex: `Change Log`.
```suggestion
## Change Log
```

Author: Sorry, you're becoming my proof-reader. Yes, that should be `Change Log`.

Reviewer: No problem, that's what a second set of eyes is for.

See [CHANGELOG.md](changelog.md)

## Meta

Troy Forster – [@tforster](https://twitter.com/tforster) – troy.forster@gmail.com

> Distributed under the ISC license. See `LICENSE` for more information.

Reviewer: Does this LICENSE file exist yet?

Author: It does not. It's probably a good habit to get into, adding an actual license file along with the simple mention of a license in package.json. It is another "control" file in the root though. I'm ok with that because it serves a unique purpose. It's not quite the same has having a katrillion dot-files all vying for some linting config attention.

What say you to adding LICENSE to our template. We could have two, one is our typical permissive OSS license (actual choice tbd) and the other we would reserve for private client projects. LICENSE-OSS.md and LICENSE-CLOSED.md and the instructions say delete one, rename the other.

Reviewer: That sounds good to me.
